.werft/util/shell.ts

```diff
@@ -62,6 +62,66 @@ export function exec(cmd: string, options?: ExecOptions): ChildProcess | shell.S
     }
 }
 
+/**
+ * Execute a command and stream output logs.
+ *
+ * If a slice is given logs are streamed using the werft log syntax; else they're streamed directly
+ * to stderr/stdout.
+ */
+export async function execStream(command: string, options: ExecOptions ): Promise<ExecResult> {
+    const werft = getGlobalWerftInstance();
+
+    options = options || {};
+
+    if (options.slice) {
+        options.silent = true;
+    }
+
+    const child = shell.exec(command, {...options, async: true});
+
+    let stdout = '';
+    let stderr = '';
+
+    // note: the stdout/stderr event handlers aren't guaranteed to receive buffers that are always
+    // newline terminated. The original log messages can be preserved by finding the index of the
+    // last newline, printing up until that newline, buffering the remaining message, appending
+    // to that buffer on the next call, and finally flushing the buffers then the process exits.
+    child.stdout.on('data', (data) => {
+        if (options.slice) werft.logOutput(options.slice, data.trim());
+        stdout += data;
+    });
+
+    child.stderr.on('data', (data) => {
+        if (options.slice) werft.logOutput(options.slice, data.trim());
+        stderr += data;
+    });
+
+    const code = await new Promise<number>((resolve, reject) => {
+        child.on('close', (code) => {
+            if (typeof code === "number") {
```

Comment:
I don't think we can separate these three if/else if/else conditions

If the exit code is non-zero I would expect it to `reject(code)` _unless_ `options.dontCheckRc` is set ☺️

Reply:
Good catch, thanks! Also, I misread the docs; `on('close')` passes both the [exit code and signal](https://github.com/nodejs/node/blob/0633e9a0b5dcbc6d3c250c9086e9386c5d5261b3/lib/internal/child_process.js#L1098) as separate values, which simplifies this quite a bit.